src/elements/text-input/src/index.tsx

```diff
@@ -1,44 +1,91 @@
-import * as React from 'react'
+/** @jsx jsx */
 
-import { DataProps, Input, Width } from './common'
+import { useState } from 'react'
+import { jsx } from '@emotion/core'
+import { FrozenInput } from '@uswitch/trustyle.frozen-input'
+import { inputs } from '@uswitch/trustyle.styles'
+import * as ReactInputMask from 'react-input-mask'
 
-export interface CommonInputProps {
-  dataProps?: DataProps
+import * as st from './styles'
+
+
+const MaskedInput: any = ReactInputMask
+
+export type InputType = 'text' | 'email' | 'tel' | 'date'
+export type Width = 'half' | 'full'
+
+export interface Props extends React.InputHTMLAttributes<HTMLInputElement> {
+  defaultValue?: string | undefined
   freezable?: boolean
   hasError?: boolean
-  label?: string
+  mask?: string
   name: string
-  onBlur: () => void
-  onChange: (value: string) => void
-  onFocus?: () => void
-  placeholder?: string
-  value: string
+  prefix?: string
+  suffix?: string
+  value?: string | undefined
   width?: Width
 }
-export interface TextInputProps extends CommonInputProps {
-  mask?: string
-  maxLength?: number
-}
 
-export interface DateInputProps extends CommonInputProps {
-  maxDate?: string
-  minDate?: string
-}
+export const Input: React.FC<Props> = ({
+  freezable,
+  hasError = false,
+  mask,
+  prefix,
+  suffix,
+  width = 'full',
+  ...inputProps
+}) => {
+  const [hasFocus, setHasFocus] = useState(false)
+  const onBlur = (event: React.FocusEvent<HTMLInputElement>) => {
+    setHasFocus(false)
 
-export interface EmailInputProps extends CommonInputProps {}
+    if (inputProps.onBlur) {
+      inputProps.onBlur(event)
+    }
+  }
+  const onFocus = (event: React.FocusEvent<HTMLInputElement>) => {
+    setHasFocus(true)
 
-export const TextInput: React.FC<TextInputProps> = props => (
-  <Input {...props} type="text" />
-)
+    if (inputProps.onFocus) {
+      inputProps.onFocus(event)
+    }
+  }
+
+  const childProps = {
+    ...inputProps,
+    css: inputs.keyboardInput,
+    onBlur,
+    onFocus
+  }
+
+  return (
+    <FrozenInput
+      text={inputProps.value || inputProps.defaultValue || ''}
+      freezable={freezable}
+    >
+      <div css={[inputs.keyboardInputContainer(hasError, hasFocus), st[width]]}>
+        {prefix && <span css={st.prefix(hasError, hasFocus)}>{prefix}</span>}
 
-export const DateInput: React.FC<DateInputProps> = props => (
-  <Input {...props} type="date" />
+        {mask ? (
+          <MaskedInput mask={mask} {...childProps} />
+        ) : (
+          <input {...childProps} />
+        )}
+
+        {suffix && <span css={st.suffix(hasError, hasFocus)}>{suffix}</span>}
+      </div>
+    </FrozenInput>
+  )
+}
+
+export const TextInput: React.FC<Props> = props => (
+  <Input {...props} type="text" />
```

Comment:
Could `type` not just be validated as one of `[text, email, tel, ...]` rather than exporting several other components? I don't see the value of this.

Reply:
Without having separate props, no real value. I thought it might be visually helpful, but since consumers are probably going to wrap it in something else anyway, before using it on a page, it's probably not worth it.

Do you have any objection to creating a fresh trustyle.input package?

Reply:
I think we should totally do that 👍